modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C

```diff
@@ -0,0 +1,89 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ComputeWeightedGapLMMechanicalContact.h"
+#include "SubProblem.h"
+
+registerMooseObject("ContactApp", ComputeWeightedGapLMMechanicalContact);
+
+InputParameters
+ComputeWeightedGapLMMechanicalContact::validParams()
+{
+  InputParameters params = ADMortarConstraint::validParams();
+  params.addClassDescription("Computes the weighted gap that will later be used to enforce the "
+                             "zero-penetration mechanical contact conditions");
+  params.addParam<NonlinearVariableName>("secondary_disp_y",
+                                         "The y displacement variable on the secondary face");
+  params.addParam<NonlinearVariableName>("primary_disp_y",
+                                         "The y displacement variable on the primary face");
+  return params;
+}
+
+ComputeWeightedGapLMMechanicalContact::ComputeWeightedGapLMMechanicalContact(
+    const InputParameters & parameters)
+  : ADMortarConstraint(parameters),
+    _secondary_disp_y(isParamValid("secondary_disp_y")
+                          ? &this->_subproblem.getStandardVariable(
+                                _tid, parameters.getMooseType("secondary_disp_y"))
+                          : nullptr),
+    _primary_disp_y(isParamValid("primary_disp_y")
+                        ? &this->_subproblem.getStandardVariable(
+                              _tid, parameters.getMooseType("primary_disp_y"))
+                        : isParamValid("secondary_disp_y")
+                              ? &this->_subproblem.getStandardVariable(
+                                    _tid, parameters.getMooseType("secondary_disp_y"))
+                              : nullptr),
+    _computing_gap_dependence(false),
+    _secondary_disp_y_sln(nullptr),
+    _primary_disp_y_sln(nullptr)
+{
+  if (_secondary_disp_y)
+  {
+    mooseAssert(
+        _primary_disp_y,
+        "It doesn't make any sense that we have a secondary displacement variable and not a "
+        "primary displacement variable");
+    _computing_gap_dependence = true;
```

Comment:
`_computing_gap_dependence` just refers to computing its Jacobian, doesn't it? When wouldn't we want to use it?

Reply:
We probably do always want to compute the (approximate) Jacobian. In fact my tests never converged if I didn't have those approximate Jacobians added in.

I copied this logic over from the other class...but yea I think what we want to do is make the `secondary_disp_y` parameter required and then just always compute the gap dependence. :+1: